python/API_GW_REQUIRE_AUTHORIZER/API_GW_REQUIRE_AUTHORIZER.py

```diff
@@ -0,0 +1,505 @@
+    """Form the evaluation(s) to be return to Config Rules
+
+    Return either:
+    None -- when no result needs to be displayed
+    a string -- either COMPLIANT, NON_COMPLIANT or NOT_APPLICABLE
+    a dictionary -- the evaluation dictionary, usually built by build_evaluation_from_config_item()
+    a list of dictionary -- a list of evaluation dictionary , usually built by build_evaluation()
+
+    Keyword arguments:
+    event -- the event variable given in the lambda handler
+    configuration_item -- the configurationItem dictionary in the invokingEvent
+    valid_rule_parameters -- the output of the evaluate_parameters() representing validated parameters of the Config Rule
+
+    Advanced Notes:
+    1 -- if a resource is deleted and generate a configuration change with ResourceDeleted status, the Boilerplate code will put a NOT_APPLICABLE on this resource automatically.
+    2 -- if a None or a list of dictionary is returned, the old evaluation(s) which are not returned in the new evaluation list are returned as NOT_APPLICABLE by the Boilerplate code
+    3 -- if None or an empty string, list or dict is returned, the Boilerplate code will put a "shadow" evaluation to feedback that the evaluation took place properly
+    """
+    config_rule_event = json.loads(event.get("invokingEvent"))
+
+    if valid_rule_parameters is not None:
+        valid_authorizers = valid_rule_parameters.get("RequireAuthTypes")
+
+    if "configurationItem" in config_rule_event:
+        configuration_item = config_rule_event.get("configurationItem")
+        api_id = configuration_item.get("resourceId").split("/")[2]
+        api_region = configuration_item.get("awsRegion")
+
+        global _client_cache
+        if _client_cache is None:
+            _client_cache = {api_region: get_client(service="apigateway", event=event, region=api_region)}
+
+        api_client = _client_cache.get(api_region)
+
+        # get all resources for this API
+        resources = api_client.get_resources(
+            restApiId=api_id, limit=500,
+            embed=[
+                'methods',
+            ]
+        )
+
+        if "items" not in resources or resources.get("items") is None or len(resources.get("items")) == 0:
+            return 'NOT_APPLICABLE'
+        else:
+            # process each resource to ensure it has an authoriser
+            methods_compliant = True
+            no_auth = 0
+            invalid_auth = 0
+            ok_auth = 0
+
+            for resource in resources.get("items"):
+                if "resourceMethods" in resource:
+                    for method, method_config in resource.get("resourceMethods").items():
+                        resource_method = f"{api_id}:{method}:{resource.get('path')}"
+
+                        auth = method_config.get("authorizationType", None)
+
+                        if auth is None or auth == 'NONE':
+                            methods_compliant = False
```

Comment:
As soon as you have a False, you should break the loop to limit API usage

Reply:
Agreed